File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include <queue>
#define MX 105
#define X first
#define Y second
using namespace std;

int arr[MX][MX];
int dx[4] = {-1,0,1,0}, dy[4] = {0,-1,0,1};

int solution(vector<vector<int>> maps)
{   
    queue<pair<int,int>> q;
    int n = maps.size(), m = maps[0].size();
    for(int i = 0; i < n; i++) fill(arr[i], arr[i]+m, -1);
    
    q.push({0,0});
    arr[0][0] = 1;
    
    while(!q.empty()){
        auto cur = q.front(); q.pop();
        int x = cur.X, y = cur.Y;
        for(int i = 0; i < 4; i++){
            int nx = x + dx[i], ny = y + dy[i];
            if(nx == n-1 && ny == m-1) return arr[x][y] + 1;
            if(nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
            if(arr[nx][ny] != -1 || maps[nx][ny] == 0) continue;
            arr[nx][ny] = arr[x][y] + 1;
            q.push({nx, ny});
        }
    }

    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int[] dx = { -1, 0, 1, 0};
    int[] dy = { 0, -1, 0, 1};
    
    public int solution(int[][] maps) {
        int n = maps.length, m = maps[0].length;
        int answer = Integer.MAX_VALUE;
        boolean[][] v = new boolean[n][m];
        
        Queue<ROR> q = new LinkedList<>();
        q.add(new ROR(0,0,0));       
        v[0][0] = true;
        
        while(!q.isEmpty()){
            ROR cur = q.poll();
            if(cur.x == n - 1 && cur.y == m - 1) {
                if(cur.cnt < answer) answer = cur.cnt + 1;
            }
            
            for(int d = 0; d < 4; d++){
                int nx = cur.x + dx[d];
                int ny = cur.y + dy[d];
                if(nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
                if(maps[nx][ny] == 0 || v[nx][ny]) continue;
                v[nx][ny] = true;
                q.add(new ROR(nx, ny, cur.cnt + 1));
            }
        }
        
        return (answer == Integer.MAX_VALUE) ? -1 : answer;
    }
}

class ROR{
    public int x; 
    public int y;
    public int cnt;
    
    ROR(int x, int y, int cnt){
        this.x = x;
        this.y = y;
        this.cnt = cnt;
    }
}